File: DLL.hs
import System.IO
import Data.List

data List = N | E Integer List deriving(Show, Eq)
-- E is a void type (Null)
-- N is a node with a integer value and the next node

-- list1: recursively defined list
newlist :: List
newlist = N

list1 :: List
list1 = E 3 $ E 8 $ E 2 $ E 6 $ E 5 $ N 

-- binder: get the internal value from the maybe monad
binder :: Maybe Integer -> Integer
binder mx = case mx of
    Just x -> x
    Nothing -> -1

-- getHead: get the first element of the list
getHead :: List -> Maybe Integer
getHead N = Nothing
getHead (E x _) = Just x

-- getTail: get the tail list of the list
getTail :: List -> List
getTail N = N
getTail (E _ l) = l

-- getLast: get the last element of the list
getLast :: List -> Maybe Integer
getLast N = Nothing
getLast (E x N) = Just x
getLast (E _ n) = getLast(n)

-- addFirst: insert an integer element at the head
addFirst :: List -> Integer -> List
addFirst N x = (E x N)
addFirst l x = (E x l)

-- delFirst: delete the first element of the list
delFirst :: List -> List
delFirst N = N
delFirst (E x l) = l

-- addLast: add the element at the last
addLast :: List -> Integer -> List
addLast N x = (E x N)
addLast l x = addFirst (addLast (delFirst l) x) (binder $ getHead l)

-- delLast: delete the last element of the list
delLast :: List -> List
delLast N = N
delLast (E x N) = N
delLast l = addFirst (delLast $ delFirst l) (binder $ getHead l)

-- findN: find N-th element of the list
findN :: List -> Integer -> Maybe Integer
findN N _ = Nothing
findN l 0 = getHead l
findN l n = findN (delFirst l) (n-1)

-- repN: replace the N-th node's value
repN :: List -> Integer -> Integer -> List
repN N _ _ = N
repN l 0 x = addFirst (delFirst l) x
repN l n x = addFirst (repN (delFirst l) (n-1) x) (binder $ getHead l)

-- addN: add a value at the N-th node (thus, nodes from N to last will be pushed by 1)
addN :: List -> Integer -> Integer -> List
addN N _ x = (E x N)
addN l 0 x = addFirst l x
addN l n x = addFirst (addN (delFirst l) (n-1) x) (binder $ getHead l)

-- delN: delete the N-th node
delN :: List -> Integer -> List
delN N _ = N
delN l 0 = delFirst l
delN l n = addFirst (delN (delFirst l) (n-1)) (binder $ getHead l)

-- toArray: convert the list in to an array
toArray :: List -> [Integer]
toArray N = []
toArray l = sort $ (binder $ getHead l) : (toArray $ getTail l)

-- toList: convert the array in to a list
toList :: [Integer] -> List
toList [] = N
toList xs = addFirst (toList $ tail xs) (head xs)

-- sortList: sort the list
sortList :: List -> List
sortList l = (toList.toArray) l

-- trv: traversing the list
trv :: List -> IO ()
trv N = promptLn ""
trv l = do
    prompt $ show $ binder $ getHead l
    prompt " "
    trv $ delFirst l

prompt :: String -> IO ()
prompt str = (putStr str)
    >>= (\x -> hFlush stdout)

promptLn :: String -> IO ()
promptLn str = (putStrLn str)
    >>= (\x -> hFlush stdout)

editList :: List -> IO ()
editList l = do
    prompt ">> enter a command: "
    cmd <- getLine
    case cmd of 
        "exit"      -> do
            prompt ">> final list: "
            trv l

        "traverse"  -> trv l >>= (\x -> editList l)

        "reset"     -> do
            promptLn ">> are you sure? [Y]/[N]"
            prompt ">> "
            yn <- getLine
            case yn of 
                "Y"     -> do
                    promptLn "resetting the list..."
                    editList newlist
                "yes"   -> do
                    promptLn "resetting the list..."
                    editList newlist
                _       -> do
                    editList l

        "--help"    -> do
            promptLn "--------------------------------------------------------------------"
            promptLn "  exit                - exit the list editing"
            promptLn "  traverse            - traverse the current list"
            promptLn "  reset               - reset the current list to an empty list"
            promptLn "  append              - add an integer at the last of the list"
            promptLn "  pop                 - delete the last element of the list"
            promptLn "  enqueue             - add an integer at the first of the list"
            promptLn "  dequeue             - delete the first element of the list"
            promptLn "  replace             - replace an element in the list"
            promptLn "  insert              - insert an element at the desired index"
            promptLn "  delete              - delete an element at the desired index"
            promptLn "  sort                - sort the list"
            promptLn "--------------------------------------------------------------------"
            editList l

        "append"    -> do
            prompt ">> enter an integer to append: "
            cx <- getLine
            let x = (read cx :: Integer)
            trv $ addLast l x
            editList $ addLast l x

        "pop"       -> do
            promptLn ">> last element popped"
            trv $ delLast l
            editList $ delLast l

        "enqueue"   -> do
            prompt ">> enter an integer to enqueue: "
            cx <- getLine
            let x = (read cx :: Integer)
            trv $ addFirst l x
            editList $ addFirst l x

        "dequeue"   -> do
            promptLn ">> first element dequeued"
            trv $ delFirst l
            editList $ delFirst l

        "insert"    -> do
            prompt ">> enter the index to add an element: "
            cn <- getLine
            prompt $ ">> enter the integer to be added at postion [" ++ cn ++ "]: "
            cx <- getLine
            let n = (read cn :: Integer)
            let x = (read cx :: Integer)
            trv $ addN l n x
            editList $ addN l n x

        "sort"      -> do
            promptLn ">> sorting the list..."
            trv $ sortList l
            editList $ sortList l

        "replace"   -> do
            prompt ">> enter the index to replace an element: "
            cn <- getLine
            prompt $ ">> enter the integer to replace at postion [" ++ cn ++ "]: "
            cx <- getLine
            let n = (read cn :: Integer)
            let x = (read cx :: Integer)
            trv $ repN l n x
            editList $ repN l n x
            
        "delete"    -> do
            prompt ">> enter the index to delete an element: "
            cn <- getLine
            let n = (read cn :: Integer)
            trv $ delN l n
            editList $ delN l n
            
        _           -> do
            promptLn $ ">> no such command <" ++ cmd ++ ">"
            promptLn $ ">> use --help to see the list of the commands"
            editList l
    
main :: IO ()
main = do
    promptLn "--------------------------------------------------------------------"
    promptLn "Dynamic List in Haskell"
    editList newlist
    promptLn "--------------------------------------------------------------------"
    return()

File: LinkedList/DLL.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>  

struct Node {
    int val;
    struct Node* prev;
    struct Node* next;
};

struct List {
    int len;
    struct Node* head;
    struct Node* tail;
};

struct List* construct() {
    struct List* newList = (struct List*)malloc(sizeof(struct List));
    newList->len = 0;
    newList->head = NULL;
    newList->tail = NULL;
    return newList;
}

void destruct(struct List* list) {
    free(list);
}

void append(struct List* list, int v) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->next = NULL;
    node->val = v;
    node->prev = list->tail;
    if (list->tail != NULL) {
        list->tail->next = node;
    }
    else list->head = node;
    list->tail = node;
    (list->len)++;
}

void enqueue(struct List* list, int v) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->prev = NULL;
    node->val = v;
    node->next = list->head;
    if (list->head != NULL) {
        list->head->prev = node;
    }
    else list->tail = node;
    list->head = node;
    (list->len)++;
}

int pop(struct List* list) {
    if (list->head == NULL) return 1<<31;   // error code
    int v = list->tail->val;
    struct Node* ptail = list->tail->prev;
    free(list->tail);
    if (ptail != NULL) {
        list->tail = ptail;
        ptail->next = NULL;
    }
    else {
        list->head = NULL;
        list->tail = NULL;
    }
    list->len -= 1;
    return v;
}

int dequeue(struct List* list) {
    if (list->head == NULL) return 1<<31;
    int v = list->head->val;
    struct Node* nhead = list->head->next;
    free(list->head);
    if (nhead != NULL) {
        list->head = nhead;
        nhead->prev = NULL;
    }
    else {
        list->head = NULL;
        list->tail = NULL;
    }
    list->len -= 1;
    return v;
}

void replaceN(struct List* list, int index, int v) {
    if (index >= list->len) return;     // out of index
    else { 
        struct Node* walk = list->head;
        for (int i=0; i<index; i++) {
            walk = walk->next;
        }
        walk->val = v;
    }
}

void insertN(struct List* list, int index, int v) {
    if (index == 0) {
        enqueue(list, v);
        return;
    }
    if (index == (list->len-1)) {
        append(list, v);
        return;
    } 
    if (index >= list->len) return;     // out of index
    else {
        struct Node* walk = list->head;     
        struct Node* node = (struct Node*)malloc(sizeof(struct Node));
        node->val = v;
        for (int i=0; i<index; i++) {
            walk = walk->next;
        }
        walk->prev->next = node;
        node->next = walk;
        node->prev = walk->prev;
        walk->prev = node;
        list->len += 1;
    }
}

void deleteN (struct List* list, int index) {
    if (index == 0) {
        dequeue(list);
        return;
    }
    if (index == list->len-1) {
        pop(list);
        return;
    }
    if (index >= list->len) return;     // out of index
    else {
        struct Node* walk = list->head;
        for (int i=0; i<index; i++) {
            walk = walk->next;
        }
        walk->prev->next = walk->next;
        walk->next->prev = walk->prev;
        free(walk);
        list->len -= 1;
    }
}

void insSort(int *arr, int size) {
    int temp = 0;
    for (int i = 1; i < size; i++) {
        for (int j = 0; j < i; j++) {
            if (*(arr+i) < *(arr+j)) {
                temp = *(arr+i);
                *(arr+i) = *(arr+j);
                *(arr+j) = temp;
            }
        }
    }
}

void sort(struct List* list) {
    int l = list->len;
    int* array = (int*)malloc(l * sizeof(int));
    struct Node* walk = list->head;
    for (int i=0; i<l; i++) {
        *(array + i) = walk->val;
        walk = walk -> next;
    }
    insSort(array, l);
    struct Node* temp = list->head;
    for (int i=0; i<l; i++) {
        temp->val = *(array + i);
        temp = temp->next;
    }
    free(array);
}

void traverse(struct List* list) {
    struct Node* walk = list->head;
    while(walk != NULL) {
        printf("%d ", walk->val);
        walk = walk->next;
    }
    printf("\n");
}

int main() {
    struct List* list1 = construct();
    printf("--------------------------------------------------------------------\n");
    printf("Dynamic List in C\n");
    int status = 1;
    while (status) {
        char cmd[20];
        printf(">> enter a command: ");
        scanf("%19s", cmd);
        if (!strcmp(cmd, "exit")) {
            destruct(list1);
            status = 0;
        }
        else {
            if (!strcmp(cmd, "traverse")) {
                traverse(list1);
            }
            int val = 0;
            int ind = 0;
            int stt = 1;
            if (!strcmp(cmd, "append") || !strcmp(cmd, "A")) {
                printf(">> enter an integer to append: ");
                scanf("%d", &val);
                append(list1, val);
            }
            else if (!strcmp(cmd, "pop") || !strcmp(cmd, "P")) {
                printf(">> last element popped\n");
                pop(list1);
            }
            else if (!strcmp(cmd, "enqueue") || !strcmp(cmd, "E")) {
                printf(">> enter an integer to enqueue: ");
                scanf("%d", &val);
                enqueue(list1, val);
            }
            else if (!strcmp(cmd, "dequeue") || !strcmp(cmd, "D")) {
                printf(">> first element dequeued\n");
                dequeue(list1);
            }
            else if (!strcmp(cmd, "replace") || !strcmp(cmd, "R")) {
                printf(">> enter the index to replace an element: ");
                scanf("%d", &ind);
                printf(">> enter the integer to replace at postion [%d]: ", ind);
                scanf("%d", &val);
                replaceN(list1, ind, val);
            }
            else if (!strcmp(cmd, "insert") || !strcmp(cmd, "I")) {
                printf(">> enter the index to insert an element: ");
                scanf("%d", &ind);
                printf(">> enter the integer to insert at postion [%d]: ", ind);
                scanf("%d", &val);
                insertN(list1, ind, val);
            }
            else if (!strcmp(cmd, "delete") || !strcmp(cmd, "X")) {
                printf(">> enter the index to delete an element: ");
                scanf("%d", &ind);
                deleteN(list1, ind);
            }
            else if (!strcmp(cmd, "sort") || !strcmp(cmd, "S")) {
                printf(">> sorting the list...\n");
                sort(list1);
            }
            else if (!strcmp(cmd, "reset")) {
                printf(">> resetting the list...\n");
                free(list1);
                struct List* list1 = construct();
            }
            else if (!strcmp(cmd, "--help")) {
                printf("--------------------------------------------------------------------\n");
                printf("    exit                - exit the list editing\n");
                printf("    traverse            - traverse the current list\n");
                printf("    reset               - reset the current list to an empty list\n");
                printf("    append  [A]         - add an integer at the last of the list\n");
                printf("    pop     [P]         - delete the last element of the list\n");
                printf("    enqueue [E]         - add an integer at the first of the list\n");
                printf("    dequeue [D]         - delete the first element of the list\n");
                printf("    replace [R]         - replace an element in the list\n");
                printf("    insert  [I]         - insert an element at the desired index\n");
                printf("    delete  [X]         - delete an element at the desired index\n");
                printf("    sort    [S]         - sorts the list in ascending order\n");
                printf("--------------------------------------------------------------------\n");
                stt = 0;
            }
            else {
                stt = 0;
                printf(">> no such command <%s>\n", cmd);
                printf(">> use --help to see the list of the commands\n");
            }
            if (stt) traverse(list1);
        }
    }
    printf("--------------------------------------------------------------------\n");
    return 0;
}
